Makefile-v2.mk

```diff
@@ -0,0 +1,296 @@
+endif
+
+#
+# Targets for building Docker images
+#
+
+DOCKER_FLAGS := --build-arg RELEASE_BUILD=$(RELEASE_BUILD) --build-arg VERSION=$(VERSION)
+
+.PHONY: images agent-image agentctl-image operator-image crow-image smoke-image
+images: agent-image agentctl-image operator-image crow-image smoke-image
+
+agent-image:
+	DOCKER_BUILDKIT=1 docker build $(DOCKER_FLAGS) -t $(AGENT_IMAGE) -f cmd/agent/Dockerfile.new .
+agentctl-image:
+	DOCKER_BUILDKIT=1 docker build $(DOCKER_FLAGS) -t $(AGENTCTL_IMAGE) -f cmd/agentctl/Dockerfile.new .
+operator-image:
+	DOCKER_BUILDKIT=1 docker build $(DOCKER_FLAGS) -t $(OPERATOR_IMAGE) -f cmd/agent-operator/Dockerfile.new .
+crow-image:
+	DOCKER_BUILDKIT=1 docker build $(DOCKER_FLAGS) -t $(CROW_IMAGE) -f tools/crow/Dockerfile.new .
+smoke-image:
+	DOCKER_BUILDKIT=1 docker build $(DOCKER_FLAGS) -t $(SMOKE_IMAGE) -f tools/smoke/Dockerfile.new .
+
+#
+# Targets for generating assets
+#
+
+.PHONY: generate generate-crds generate-manifests generate-dashboards generate-protos
+generate: generate-crds generate-manifests generate-dashboards generate-protos
+
+generate-crds:
+ifeq ($(USE_CONTAINER),1)
+	$(RERUN_IN_CONTAINER)
+else
+	bash ./tools/generate-crds.bash
+endif
+
+generate-manifests:
+ifeq ($(USE_CONTAINER),1)
+	$(RERUN_IN_CONTAINER)
+else
+	cd production/kubernetes/build && bash build.sh
+endif
+
+generate-dashboards:
+ifeq ($(USE_CONTAINER),1)
+	$(RERUN_IN_CONTAINER)
+else
+	cd example/docker-compose && jb install && \
+	cd grafana/dashboards && jsonnet template.jsonnet -J ../../vendor -m .
+endif
+
+generate-protos:
+ifeq ($(USE_CONTAINER),1)
+	$(RERUN_IN_CONTAINER)
+else
+	go generate ./pkg/agentproto/
+endif
+
+#
+# Other targets
+#
+# build-container-cache and clean-build-container-cache are defined in
+# Makefile.build-container.
+#
+
+.PHONY: clean
+clean: clean-dist clean-build-container-cache
+	rm -rf ./build/*
+
+.PHONY: info
+info:
+	@printf "USE_CONTAINER   = $(USE_CONTAINER)\n"
+	@printf "AGENT_IMAGE     = $(AGENT_IMAGE)\n"
+	@printf "AGENTCTL_IMAGE  = $(AGENTCTL_IMAGE)\n"
+	@printf "OPERATOR_IMAGE  = $(OPERATOR_IMAGE)\n"
+	@printf "CROW_IMAGE      = $(CROW_IMAGE)\n"
+	@printf "SMOKE_IMAGE     = $(SMOKE_IMAGE)\n"
+	@printf "BUILD_IMAGE     = $(BUILD_IMAGE)\n"
+	@printf "AGENT_BINARY    = $(AGENT_BINARY)\n"
+	@printf "AGENTCTL_BINARY = $(AGENTCTL_BINARY)\n"
+	@printf "OPERATOR_BINARY = $(OPERATOR_BINARY)\n"
+	@printf "CROW_BINARY     = $(CROW_BINARY)\n"
+	@printf "SMOKE_BINARY    = $(SMOKE_BINARY)\n"
+	@printf "GOOS            = $(GOOS)\n"
+	@printf "GOARCH          = $(GOARCH)\n"
+	@printf "GOARM           = $(GOARM)\n"
+	@printf "CGO_ENABLED     = $(CGO_ENABLED)\n"
+	@printf "RELEASE_BUILD   = $(RELEASE_BUILD)\n"
+	@printf "VERSION         = $(VERSION)\n"
+	@printf "GO_TAGS         = $(GO_TAGS)\n"
+
+# awk magic to print out the comment block at the top of this file.
+.PHONY: help
+help:
+	@awk 'BEGIN {FS="## "} /^##\s*(.*)/ { print $$2 }' $(MAKEFILE_LIST)
```

Comment:
Definitely like this help target 👍  I believe grafana/grafana does something similar, taken from this blog post I believe: https://www.thapaliya.com/en/writings/well-documented-makefiles/.  It seems we could take this a step further, take a look and see what you think.

Reply:
I originally started with that, but eventually moved to this because I felt it was cleaner and more flexible to have a giant doc comment at the top of the Makefile, since that means it's also more accessible to people reading the Makefile directly.